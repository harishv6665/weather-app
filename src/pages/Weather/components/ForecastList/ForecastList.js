import React, {useEffect, useState} from "react";
import PropTypes from "prop-types";
import ForecastCard from "../ForecastCard/ForecastCard";
import styles from "./ForecastList.module.css";
import {Button, Switch} from "../../../../shared/components";

const ForecastList = ({ list, activeForecastDate, setActiveForecastDate, isCelsius, setIsCelsius, city }) => {
    const [data, setData] = useState([]);
    const [currentPage, setCurrentPage] = useState(1);
    const [pageSize] = useState(3);
    const [totalPages, setTotalPages] = useState(0);

    useEffect(() => {
        setData(list);
        setTotalPages(Math.ceil(list.length / pageSize));
        setActiveForecastDate(list[0].date);
    }, [list]);

    const getPageData = () => {
        const endIndex = currentPage * pageSize;
        const startIndex = endIndex - pageSize;
        return data.slice(startIndex, endIndex);
    }

    const onPrevPage = () => {
        if (currentPage > 1) {
            setCurrentPage(currentPage - 1);
        }
    }

    const onNextPage = () => {
        if (totalPages > currentPage) {
            setCurrentPage(currentPage + 1);
        }
    }

    return (
        <div className={styles.forecast}>
            <div className={styles.forecast__header}>
                <h2 className={styles.forecast__title}>{city}</h2>
                <div className={styles.forecast__buttonGroup}>
                    <label className={styles.forecast__switch}>
                        <span className={styles.forecast__switch__label}>ºF</span>
                        <Switch
                            checked={isCelsius}
                            onChange={setIsCelsius}
                        />
                        <span className={styles.forecast__switch__label}>ºC</span>
                    </label>
                    <Button
                        onClick={() => onPrevPage()}
                        disabled={!(currentPage > 1)}
                        customStyles={{
                            button: `${styles.forecast__button}
                            ${styles.forecast__buttonPrev}`,
                        }}
                    />
                    <Button
                        onClick={() => onNextPage()}
                        disabled={!(totalPages > currentPage)}
                        customStyles={{
                            button: `${styles.forecast__button}
                            ${styles.forecast__buttonNext}`,
                        }}
                    />
                </div>
            </div>
            <div className={styles.forecast__cards}>
                {getPageData().map(data => <ForecastCard
                    key={data.date}
                    data={data}
                    isCelsius={isCelsius}
                    isActive={data.date === activeForecastDate}
                    onCardClick={() => setActiveForecastDate(data.date)}
                />)}
            </div>
        </div>
    )
}

ForecastList.propTypes = {
    city: PropTypes.string.isRequired,
    list: PropTypes.array.isRequired,
    activeForecastDate: PropTypes.string.isRequired,
    isCelsius: PropTypes.bool.isRequired,
    setActiveForecastDate: PropTypes.func.isRequired,
    setIsCelsius: PropTypes.func.isRequired,
}

ForecastList.defaultProps = {
    city: "",
    list: [],
    activeForecastDate: "",
    isCelsius: false,
    setActiveForecastDate: () => {},
    setIsCelsius: () => {},
}

export default ForecastList;
