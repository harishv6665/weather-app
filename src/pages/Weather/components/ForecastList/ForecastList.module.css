.forecast {

}

.forecast__header {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.forecast__title {
    font-size: 18px;
    color: var(--primaryText);
    font-weight: 600;
}

.forecast__switch {
    display: flex;
    align-items: center;
    cursor: pointer;
}

.forecast__switch__label {
    font-size: 16px;
}

.forecast__buttonGroup {
    display: flex;
}

.forecast__button {
    min-width: auto;
    width: 30px;
    height: 30px;
    padding: 0;
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
    margin-left: 24px;
}

.forecast__button:before {
    content: "";
    position: absolute;
    width: 8px;
    height: 8px;
    border: 2px solid var(--white);
    border-radius: 2px;
    transform: rotate(45deg);
}

.forecast__buttonPrev:before {
    border-right-color: transparent;
    border-top-color: transparent;
    left: 12px;
}

.forecast__buttonNext:before {
    border-left-color: transparent;
    border-bottom-color: transparent;
    right: 12px;
}

.forecast__cards {
    display: flex;
    flex-direction: column;
    width: 100%;
    padding-top: 16px;
}

@media only screen and (min-width: 768px) {
    .forecast__cards{
        flex-direction: row;
        margin-right: 0;
    }

}
