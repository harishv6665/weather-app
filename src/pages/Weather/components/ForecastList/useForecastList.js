import {useEffect, useState} from "react";

const useForecastList = (list, setActiveForecastDate) => {
    const [data, setData] = useState([]);
    const [currentPage, setCurrentPage] = useState(1);
    const [pageSize] = useState(3);
    const [totalPages, setTotalPages] = useState(0);

    useEffect(() => {
        setData(list);
        setTotalPages(Math.ceil(list.length / pageSize));
        setActiveForecastDate(list[0].date);
    }, [list]);

    const getPageData = () => {
        const endIndex = currentPage * pageSize;
        const startIndex = endIndex - pageSize;
        return data.slice(startIndex, endIndex);
    }

    const onPrevPage = () => {
        if (currentPage > 1) {
            setCurrentPage(currentPage - 1);
        }
    }

    const onNextPage = () => {
        if (totalPages > currentPage) {
            setCurrentPage(currentPage + 1);
        }
    }

    return {
        getPageData,
        onPrevPage,
        onNextPage,
        totalPages,
        currentPage,
    }
}

export default useForecastList;
