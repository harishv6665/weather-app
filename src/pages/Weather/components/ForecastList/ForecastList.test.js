import React from "react";
import { shallow } from "enzyme";
import ForecastList from "./ForecastList";

const LIST_CITY = "Munich";
const LIST_LIST = [
    {date: "2020-08-15", avg_temp: 21.16},
    {date: "2020-08-16", avg_temp: 20.8425},
    {date: "2020-08-17", avg_temp: 16.92125},
    {date: "2020-08-18", avg_temp: 17.75125},
    {date: "2020-08-19", avg_temp: 17.54875},
    {date: "2020-08-20", avg_temp: 18.456666666666667},
];
const LIST_ACTIVE_FORECAST_DATE = "2020-08-19";
const LIST_IS_CELSIUS = false;
const LIST_CELSIUS_TEXT = "ºC";
const LIST_FAHRENHEIT_TEXT = "ºF";

describe("ForecastList tests", () => {

    it("should be defined", () => {
        expect(ForecastList).toBeDefined();
    });

    it("should match to snapshot", () => {
        const wrapper = shallow(<ForecastList
            city={LIST_CITY}
            list={LIST_LIST}
            activeForecastDate={LIST_ACTIVE_FORECAST_DATE}
            isCelsius={LIST_IS_CELSIUS}
            setActiveForecastDate={() => {}}
            setIsCelsius={() => {}}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should match the DOM structure", () => {
        const wrapper = shallow(<ForecastList
            city={LIST_CITY}
            list={LIST_LIST}
            activeForecastDate={LIST_ACTIVE_FORECAST_DATE}
            isCelsius={LIST_IS_CELSIUS}
            setActiveForecastDate={() => {}}
            setIsCelsius={() => {}}
        />);
        expect(wrapper.find("Switch").length).toBe(1);
        expect(wrapper.find("Button").length).toBe(2);
        expect(wrapper.find("div div h2").length).toBe(1);
        expect(wrapper.find("div div div label").length).toBe(1);
        expect(wrapper.find("div div div label span").length).toBe(2);
        expect(wrapper.find("div div div label Switch").length).toBe(1);

        expect(wrapper.find("div div h2").text()).toEqual(LIST_CITY);
        expect(wrapper.find("div div div label span").at(0).text()).toEqual(LIST_FAHRENHEIT_TEXT);
        expect(wrapper.find("div div div label span").at(1).text()).toEqual(LIST_CELSIUS_TEXT);
    });

});
