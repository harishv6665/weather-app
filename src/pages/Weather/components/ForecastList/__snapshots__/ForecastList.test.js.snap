// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastList tests should match to snapshot 1`] = `ShallowWrapper {}`;
